src/lib/ndpi_main.c

```diff
@@ -8611,26 +8620,26 @@ void ndpi_generate_options(u_int opt) {
   case 2: /* List known risks */
     {
       for(i = 1 /* Skip no risk */; i < NDPI_MAX_RISK; i++) {
-	ndpi_risk_enum r = (ndpi_risk_enum)i;
+        ndpi_risk_enum r = (ndpi_risk_enum)i;
 
-	printf("            <Option%d value=\"%u\">%s</Option%d>\n",
-	       i, i, ndpi_risk2str(r), i);
+        fprintf(options_out, "            <Option%d value=\"%u\">%s</Option%d>\n",
+                i, i, ndpi_risk2str(r), i);
       }
     }
     break;
 
   default:
-    printf("WARNING: option -a out of range\n");
+    fprintf(options_out, "%s\n", "WARNING: option -a out of range");
     break;
   }
 }
 
 /* ****************************************************** */
 
-void ndpi_dump_risks_score() {
+void ndpi_dump_risks_score(FILE *risk_out) {
   u_int i;
 
-  printf("%3s %-48s %-8s %s %-8s %-8s\n",
+  fprintf(risk_out, "%3s %-48s %-8s %s %-8s %-8s\n",
 	 "Id", "Risk", "Severity", "Score", "CliScore", "SrvScore");
 
   for(i = 1; i < NDPI_MAX_RISK; i++) {
```

Comment:
A missing `printf->fprintf(risk_out ` below

Reply:
Oh, forgot to change it.

I just saw that `ndpi_check_dga_name` uses lots of `printf()`s. But this function is mainly invoked by the dissectors. Do you think it makes sense to replace it with either `NDPI_LOG_DBG` or `NDPI_LOG_DBG2` ?

Reply:
or use `#ifdef NDPI_ENABLE_DEBUG_MESSAGES` for all `printf()`s

Reply:
It is a little bit tricky because we want to be still able to enable/disable these logs via command line parameter in `tests/dga/dga_evaluate.c`...

Reply:
I'll try to use debug printf callback and see if it works as expected.
I think that we are wasting performance doing I/O stuff in dissectors especially for TLS/QUIC  which is widely used.